include/cetl/rtti.hpp

```diff
@@ -0,0 +1,316 @@
+/// and that CV-qualifiers and references are not considered in the type comparison;
+/// that is, <tt>T</tt> and <tt>const T&</tt> are considered the same type.
+///
+/// In order to support this RTTI capability, a type must at least define a public method
+/// <tt>static constexpr cetl::type_id _get_static_type_id_() noexcept</tt> that
+/// returns the type ID (i.e., the UUID) of the type it is defined on.
+/// Types that provide said method satisfy \ref cetl::has_static_type_id.
+/// If the type is polymorphic, it should publicly inherit from \ref cetl::rtti_helper, which will provide the
+/// necessary implementation of the RTTI-related methods along with the aforementioned
+/// \c _get_static_type_id_() method.
+///
+/// RTTI-related members are named with surrounding underscores to avoid name clashes with user-defined members,
+/// including the members of DSDL types (where identifiers surrounded with underscores are reserved).
+/// The user code should not invoke such underscored methods directly but instead use the free functions defined here.
+class rtti
+{
+public:
+    /// Implementations are adivsed to use \ref cetl::rtti_helper instead of implementing this manually.
+    /// If manual implementation is preferred, then the method body should simply forward the result of this
+    /// type's static \c _get_static_type_id_ method:
+    ///
+    /// @code
+    /// return _get_static_type_id_();
+    /// @endcode
+    ///
+    /// This method should not be invoked directly; instead, use \ref cetl::get_type_id.
+    CETL_NODISCARD virtual type_id _get_polymorphic_type_id_() const noexcept = 0;
+
+    /// Implementations are adivsed to use \ref cetl::rtti_helper instead of implementing this manually.
+    /// If manual implementation is preferred, then the method body should be implemented as follows
+    /// (add \c const in the const overload):
+    ///
+    /// @code
+    /// return (id == _get_static_type_id_()) ? static_cast<void*>(this) : base::_cast_(id);
+    /// @endcode
+    ///
+    /// Where \c base is the base class of the current class that implements this method;
+    /// if there is more than one base available, all of those that implement this \ref cetl::rtti interface
+    /// should be checked in the same way one by one and the first match (non-nullptr result) should be returned.
+    ///
+    /// The objective here is to check if the current type in the hierarchy is the one being sought;
+    /// if not, delegate the call to the base class(es); the topmost class is this \c rtti type that
+    /// simply returns a null pointer, indicating that the search has returned no matches and a
+    /// safe dynamic type conversion is not possible (at least not along the current branch of the type tree).
+    ///
+    /// This method should not be invoked directly; instead, use \ref cetl::rtti_cast.
+    /// @{
+    CETL_NODISCARD virtual void* _cast_(const type_id& id) & noexcept
```

Comment:
These need to be pure-virtual. Everything in the base class has to be pure-virtual to ensure the compiler can completely elide the interface types when optimizing.

Reply:
I understand this but do note that pure virtual methods here would require special-casing in the search method when the next base to query is `cetl::rtti`. Do we need to sacrifice generality for the sake of optimization?

Reply:
It's not an optimization. It's a standard saying that we do not utilize any concrete, multiple-inheritance. Adding these methods breaks that promise. It may also expose us to edge cases with destructors and other nasties. Rather than try to reason about all of these we should just not have concrete implementation in interface classes.

Also, you don't need the search if you use virtual inheritance, right?

Reply:
Ok, I understand your point, but the search is not related to the kind of inheritance used. It is needed to traverse the tree of inheritance to find the requested type in it, starting from the most concrete type (the runtime type) and then climbing upwards each branch of the tree.